source/tutorials/module-system/module-system.md

```diff
@@ -0,0 +1,1027 @@
++      (lib.mkIf (config.map.center != null)
++        "center=\"$(geocode ${
++          lib.escapeShellArg config.map.center
++        })\"")
+     ];
+   };
+   
+```
+
+## Splitting modules: importing marker.nix
+
+The module system allows you to split your config into multiple files
+via the `imports` attribute, which can define further modules to import.
+This allows us to logically separate options and config for different
+parts.
+
+Here we create a new module `marker.nix`, where we will declare options
+for defining markers on the map
+
+```diff
+diff --git a/default.nix b/default.nix
+index 098d135..9d9f29d 100644
+--- a/default.nix
++++ b/default.nix
+@@ -1,5 +1,9 @@
+ { lib, config, ... }: {
+ 
++  imports = [
++    ./marker.nix
++  ];
++
+   options = {
+     generate.script = lib.mkOption {
+       type = lib.types.lines;
+diff --git a/marker.nix b/marker.nix
+new file mode 100644
+index 0000000..035c28d
+--- /dev/null
++++ b/marker.nix
+@@ -0,0 +1,3 @@
++{ lib, config, ... }: {
++
++}
+```
+
+## Submodule types: Declaring map.markers option
+
+One of the most useful types of the module system is the `submodule`
+type. This type allows you to define a nested module system evaluation,
+with its own options. Every value of such a type is then interpreted
+(by default) as a `config` assignment of the nested module evaluation.
+
+In this case we're defining a `map.markers` option, whose type is a list
+of submodules with a nested `location` type, allowing us to define a
+list of markers on the map, where each assignment is type checked
+according to the submodule.
+
+```diff
+diff --git a/marker.nix b/marker.nix
+index 035c28d..6a6c686 100644
+--- a/marker.nix
++++ b/marker.nix
+@@ -1,3 +1,20 @@
+-{ lib, config, ... }: {
++{ lib, config, ... }:
++let
++
++  markerType = lib.types.submodule {
++    options = {
++      location = lib.mkOption {
++        type = lib.types.nullOr lib.types.str;
++        default = null;
++      };
++    };
++  };
++in {
++
++  options = {
++    map.markers = lib.mkOption {
++      type = lib.types.listOf markerType;
++    };
++  };
+ 
+ }
+```
+
+## Single option namespace: Defining markers in generate.requestParams
+
+Since all modules in `imports` are treated the same, we can also freely
+assign an option defined in our initial module. In this case we want to
+add some request parameters derived from the `map.markers` option, so
+that markers actually show up on the map.
+
+```diff
+diff --git a/marker.nix b/marker.nix
+index 6a6c686..c2c5da2 100644
+--- a/marker.nix
++++ b/marker.nix
+@@ -17,4 +17,26 @@ in {
+     };
+   };
+ 
++  config = {
++
++    map.markers = [
++      { location = "new york"; }
++    ];
++
++    generate.requestParams = let
++      paramForMarker = marker:
++        let
++          attributes =
++            [
++              "$(geocode ${
++                lib.escapeShellArg marker.location
++              })"
++            ];
++        in "markers=${
```

Comment:
It's kind of hard to understand what's happening here on a language level. Are you joining each attribute with the `\\|` string?